### PR TITLE
Check for require when binding commonjs

### LINKED ISSUE #48726: When using `require()` in es modules, no TS2441 error occurs and `node_modules/@types` is automatically included in the project.
# Bug Report

### 🔎 Search Terms

cjs require

### 🕗 Version & Regression Information

- This changed between versions `4.0.5` and `4.2.3`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.2&filetype=js#code/PQKhCgAIUgBAHAhgJ0QW0gbwM4BdkCWAdgOYC+keyUIw4AZgK5EDGuBA9kZMgKYCOjAnwAUVAJRYy4cCy55KHNLwCyHACaMANr0gBeHgKGiARCQ4cSOgLQstHbIz62l8AjuQnxAbhm8AHvAcyLhY2EqqGtq8ZL5AA+QA)

### 💻 Code

```ts repro
// @filename: index.js
// @checkJs: true

/**
 * @param {string} str
 */
function require(str) {}

const someModule = require("google-closure-compiler");

export {someModule};
```

### 🙁 Actual behavior

TS2792 occurs at the call to `require()`:

```js
const someModule = require("google-closure-compiler");
//                         ^^^^^^^^^^^^^^^^^^^^^^^^^----- Cannot find module 'google-closure-compiler'. option?
//                                                        Did you mean to set the 'moduleResolution' option to 'node',
//                                                        or to add aliases to the 'paths'
```

What's worse, if the project contains types in `node_modules/@types`, these will now be included in the project, regardless of the value set for `"typeRoots": []` in `tsconfig.json`, which is supposed to prevent this sort of behaviour.

From https://www.typescriptlang.org/tsconfig#typeRoots:
> If `typeRoots` is specified, only packages under `typeRoots` will be included.

For me this caused a long and painful process of figuring out why I kept getting errors for imports missing from the "path" module, even though I definitely disabled node imports.

[Here you can download](https://github.com/microsoft/TypeScript/files/8498452/jsRequireBug.zip) a reduced test case of the situation described above.

Run `tsc --noEmit -p jsconfig.json` in the folder and you'll get an error stating that `fromFileUrl` is not in the `"path"` module, even though `node_modules/@types` has been disabled and a custom path has been added in `jsconfig.json`.

### 🙂 Expected behavior

Either no error, or an error informing me that I can't use `require()` as it's already taken by the compiler. The latter is already the case with TypeScript files:
```ts
function require(str: string) {}
//       ^^^^^^^------ TS2441: Duplicate identifier 'require'.
//                     Compiler reserves name 'require' in top level scope of a module.
```
It's just missing inside JavaScript files.

### 🔃 Workarounds

Just don't name your functions `require` 😅. But you'd have to know that this issue is happening in the first place, which is not easy to figure out with no errors telling you this.

### PR DESCRIPTION
**Pull request by sandersn**
Previously, the binder didn't check for a declared function named 'require' when binding require in JS. It unconditionally treated 'require' as the commonjs function, even when a local function named 'require' existed and even in ES modules. Now, the binder checks for a local function named require and avoids binding in that case.

Both before and after, actually importing the module causes the checker to issue an error saying that top-level functions named 'require' are illegal. But this error doesn't show up until you've imported from the module, so you won't notice it in the editor, where JS errors are most useful.

Within-binder checks for declared functions are dependent on binding order, but function binding simulates hoisting by binding all function declarations first, so it should be pretty reliable.

Fixes #48726

### CODE REVIEW

**Comment by sandersn** [comment]
Note: The "'require' is reserved by the compiler" error has to do with emit, so it shouldn't be issued on any .js file. That's a separate issue though.

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -3260,6 +3260,7 @@ namespace ts {
                 if (isInJSFile(node) &&
                     isVariableDeclarationInitializedToBareOrAccessedRequire(possibleVariableDecl) &&
                     !getJSDocTypeTag(node) &&
+                    !(lookupSymbolForName(container, "require" as __String)) &&
```
Comment (ProKashif) [inline src/compiler/binder.ts:3263]: we should move this in enum as we are doing in below

**Review by ProKashif — APPROVED** [review]

File: src/compiler/binder.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/binder.ts:3263]: Are you talking about an enum like ModifierFlags? `require` isn't a modifier like `public` et al, and I think if the binder gets this right, then the checker won't have to look up `require` again. Even if it does, it'll only happen once more. I don't think it's worth caching.

**Comment by jakebailey** [comment]
Will this end up breaking code that relied on the fact that we treat a function named "require" specially? I know that yarn plugins are passed a require function as an argument (so you can require things lazily / get the same dep as the executing yarn version), and having `require` recognized as the require function is helpful there.

**Comment by andrewbranch** [comment]
Similar question to Jake’s: does this check affect ambient declarations? Both @types/node and @types/webpack-env declare `require`.

**Comment by sandersn** [comment]
@andrewbranch No: `lookupForSymbolName` doesn't look in the global scope, so it won't affect files that reference those types. However, a `declare function require()` in the same file would block `require` use there, even though that's not legal JS syntax.

@jakebailey Yes: based on the test noCrashOnParameterNamedRequire. While technically correct, the assumption that `require` behaves like commonjs require is really useful for understanding code most of the time.

ES modules would be less likely to have a `require` passed to them, I think. Do you think this change makes sense if restricted to ES modules? I went ahead and did that, but if you can access `require` in an ES module (and if people commonly do), then even that probably doesn't make sense.

I looked at a couple of yarn plugins but they were written in TS, and I didn't see any `require` passed around.

Two more things:
1. This is a common pattern in bundled code too, but the references are to other regions in the same bundled file, so we've never been able to resolve them unless the bundle is in the same directory as the source.
2. I can't figure out how to trigger the "please don't name your function `require` error in the editor." Improving the usability of that seems to be a better alternative fix at this point.

**Comment by jakebailey** [comment]
> I looked at a couple of yarn plugins but they were written in TS, and I didn't see any require passed around.

My example is in a private repo so I can't link it here, but here is the example from the yarn docs: https://yarnpkg.com/advanced/plugin-tutorial/#writing-our-first-plugin

**Comment by sandersn** [comment]
Closing this after various discussions. I think a warning against writing a function named `require` is better.

**Pull request closed without merging**